07_IPC/03_Message_Passing.md

```diff
@@ -112,72 +133,83 @@ if (target == NULL)
+You may want to return an error here if the endpoint can't be found, however, in our case we're simply discarding the message.
 
-Now we'll need to allocate a buffer to store a copy of the message in, and copy the original message into this buffer.
+Now we'll need to allocate a buffer to store a copy of the message in and copy the original message into this buffer.
 
 ```c
-void* msg_copy = malloc(length);
-memcpy(msg_copy, buffer, length);
+void* kernel_copy = kmalloc(length);
+memcpy(kernel_copy, buffer, length);
 ```
 
-Why do we make a copy of the original message? Well if we don't, the sending process has to keep the original message around until the receiver has processed it. We don't have a way for the receiver to communicate that it's finished reading the message. By making a copy, we can return from `ipc_send()` as soon as the message is sent, regardless of when the message is read. Now the sending process is free to do what it wants with the memory holding the original message as soon as `ipc_send()` has completed.
+Why do we make a copy of the original message? Well if we don't, the sending process has to keep the original message around until the receiver has processed it. We don't have a way for the receiver to communicate that it's finished reading the message. By making a copy, we can return from `ipc_send()` as soon as the message is sent, regardless of when the message is read. Now the sending process is free to do what it wants with the memory holding the original message as soon as `ipc_send()` has been completed.
 
 If you're performing this IPC as part of a system call from userspace, the memory containing the original message is unlikely to be mapped in the receiver's address space anyway, so we have to copy it into the kernel's address space, which is mapped in both processes.
 
 All that's left is to tell the receiver it has a message available by placing the buffer address on the endpoint. Again, notice the use of the lock to prevent race conditions while we mess with the internals of the endpoint.
 
 ```c
 acquire(&target->lock);
-target->msg_buffer = msg_copy;
-target->msg_length = length;
-release(&target->lock);
-```
 
-After the lock on the endpoint is released, the message has been sent! Now it's up to the receiving thread to check the endpoint and set the buffer to NULL again.
+load_address_space(target->owner_pid)
 
-### Multiple Messages
+// Create the message
+ipc_message_t* new_message = (ipc_message_t*)malloc(sizeof(ipc_message_t));
+void* new_buffer = malloc(size);
+new_message -> message_buffer = memcpy(new_buffer, kernel_copy, size);
+new_message -> message_size = size;
+new_message -> next_message = 0;
 
-In theory this works, but we've overlooked one huge issue: what if there's already a message at the endpoint? You should handle this, and there's a couple of ways to go about it:
+// Add  the message to the queue
+// Left for the reader to do, trivial linked list appending
```

Comment:
this reads like a note I'd leave for myself rather than a hint. What about:
> // append message struct to the endpoint's queue, implementation is left as an exercise for the reader.